src/installed.rs

```diff
@@ -0,0 +1,357 @@
+                                             tokens.error_description
+                                                   .unwrap_or("".to_string()))
+                                         .as_str());
+            Result::Err(Box::new(err))
+        }
+    }
+
+    /// Obtains an authorization code either interactively or via HTTP redirect (see
+    /// InstalledFlowReturnMethod).
+    /// TODO(dermesser): Add timeout configurability!
+    fn get_authorization_code<'a, AD: AuthenticatorDelegate, S, T>(&mut self,
+                                                                   auth_delegate: &mut AD,
+                                                                   appsecret: &ApplicationSecret,
+                                                                   scopes: S)
+                                                                   -> Result<String, Box<Error>>
+        where T: AsRef<str> + 'a,
+              S: Iterator<Item = &'a T>
+    {
+        let result: Result<String, Box<Error>> = match self.server {
+            None => {
+                let url = build_authentication_request_url(&appsecret.auth_uri,
+                                                           &appsecret.client_id,
+                                                           scopes,
+                                                           None);
+                match auth_delegate.present_user_url(&url, true /* need_code */) {
+                    None => {
+                        Result::Err(Box::new(io::Error::new(io::ErrorKind::UnexpectedEof,
+                                                            "couldn't read code")))
+                    }
+                    // Remove newline
+                    Some(mut code) => {
+                        code.pop();
+                        Result::Ok(code)
+                    }
+                }
+            }
+            Some(_) => {
+                // The redirect URI must be this very localhost URL, otherwise Google refuses
+                // authorization.
+                let url = build_authentication_request_url(&appsecret.auth_uri,
+                                                           &appsecret.client_id,
+                                                           scopes,
+                                                           Some(format!("http://localhost:{}",
+                                                                        self.port
+                                                                            .unwrap_or(8080))));
+                auth_delegate.present_user_url(&url, false /* need_code */);
+
+                match self.auth_code_rcv.as_ref().unwrap().recv() {
+                    Result::Err(e) => Result::Err(Box::new(e)),
+                    Result::Ok(s) => Result::Ok(s),
+                }
+            }
+        };
+        let _ = self.server.as_mut().map(|l| l.close());
```

Comment:
well, `ok()` doesn't work because it's already returning an option. But `is_some()` suppresses the error too (I guess that's what your intention is)